### PR TITLE
Implementation Section

### PR DESCRIPTION
**Pull request by craftamap**

### CODE REVIEW

File: chapters/implementation.tex [hunk 1]
```diff
@@ -1,3 +1,54 @@
 \section{Implementation} \label{sec:implementation}
-Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren.
 
+For the implementation of the app, Kotlin was used.
```
Comment (mac641) [inline chapters/implementation.tex]: I'd write `The app/soHappy is built using/with Kotlin.`

File: chapters/implementation.tex [hunk 2]
```diff
@@ -1,3 +1,54 @@
...
+
+Kotlin is ``[...] an open-source statically typed programming language
+that targets the JVM, Android, JavaScript and Native.'' \cite{kotlin2020}
+Google officially supports Kotlin for Android Apps since 2017
```
Comment (limkokyew) [inline chapters/implementation.tex]: apps lowercase

File: chapters/implementation.tex [hunk 3]
```diff
@@ -1,3 +1,54 @@
...
+\cite{androidkotlin2019}.
+
+Thanks to the similarity with languages like Java and TypeScript, adapting
+Kotlin as programming language was relatively easy for the project team.
+
+Various language features of Kotlin helped in development, too. Instead of
+implementing the singleton pattern, Kotlin allows to create an single object
+per runtime by using the \texttt{object}-Keyword.
+
+Some design choices of the app were slowed down by Kotlins language features.
```
Comment (limkokyew) [inline chapters/implementation.tex]: Kotlin's

File: chapters/implementation.tex [hunk 4]
```diff
@@ -1,3 +1,54 @@
...
+As all Kotlin classes inherit from \texttt{Any}, using features of Java's 
+\texttt{Object}-class are not natively available. This made using the methods
+\texttt{.wait()} and \texttt{.notify()}, which are used for synchronisation,
+more difficult.
+
+Using ``Android Studio'', the official IDE for Android development, allowed
```
Comment (limkokyew) [inline chapters/implementation.tex]: Maybe write "integrated development environment (IDE)"

File: chapters/implementation.tex [hunk 5]
```diff
@@ -1,3 +1,54 @@
...
+build an apk for each architecture. As a minimum SDK version, version 23
+was chosen. This means that the app works on Android 6 and above.
+
+In order to improve the code quality, a linting software called ``detekt'',
+``[...] a static code analysis tool [...]''\cite{detekt2020} was used.
+
+Various libraries were used to help implement the application. Using
+the included build tool ``Gradle'' \cite{gradle} allowed easy management of 
+dependencies. The following libraries are worth mentioning: CameraX 
+\cite{camerax} is Androids newest Library allowing access to the Camera and
```
Comment (limkokyew) [inline chapters/implementation.tex]: library lowercase
camera lowercase

File: chapters/implementation.tex [hunk 6]
```diff
@@ -1,3 +1,54 @@
...
+
+In order to improve the code quality, a linting software called ``detekt'',
+``[...] a static code analysis tool [...]''\cite{detekt2020} was used.
+
+Various libraries were used to help implement the application. Using
+the included build tool ``Gradle'' \cite{gradle} allowed easy management of 
+dependencies. The following libraries are worth mentioning: CameraX 
+\cite{camerax} is Androids newest Library allowing access to the Camera and 
+providing an easy way to analyse captured images. ``GPUImage'' \cite{gpuimage}
+allowed to set filters onto the preview image. A library called ``Nitrite''
```
Comment (limkokyew) [inline chapters/implementation.tex]: I'd prefer "apply" instead of "set"

File: chapters/implementation.tex [hunk 7]
```diff
@@ -1,3 +1,54 @@
...
+\cite{nitrite} allowed us to implement an NoSQL-Database 
+storing our measurement locally.
+
+For image analysis, we used the following libraries: For cascade-based face
+detection, the official `OpenCV4Android'-Library was used\cite{opencv4android}.
+For machine learning based approaches, the official TFLite library was utilised.
+\cite{tensorflow}
+
+In order to train the CNN, which is explained in \ref{sec:methodology}, an
```
Comment (limkokyew) [inline chapters/implementation.tex]: a Python 3 script

File: chapters/implementation.tex (hunk 4 above)
Comment (mac641) [inline chapters/implementation.tex]: I'd write `Android Studio, known as the official IDE for Android development, allowed ...`

File: chapters/implementation.tex [hunk 8]
```diff
@@ -1,3 +1,54 @@
...
+\cite{nitrite} allowed us to implement an NoSQL-Database 
+storing our measurement locally.
+
+For image analysis, we used the following libraries: For cascade-based face
+detection, the official `OpenCV4Android'-Library was used\cite{opencv4android}.
+For machine learning based approaches, the official TFLite library was utilised.
+\cite{tensorflow}
+
+In order to train the CNN, which is explained in \ref{sec:methodology}, an 
+Python 3-Script running in an jupyter notebook was utilised.
```
Comment (limkokyew) [inline chapters/implementation.tex]: a Jupyter notebook

**Review by AnnaGsell — APPROVED** [review]

File: chapters/implementation.tex (hunk 6 above)
Comment (mac641) [inline chapters/implementation.tex]: Keep the writing tense consistent:
allowed -> allows

File: chapters/implementation.tex [hunk 9]
```diff
@@ -1,3 +1,54 @@
...
+In order to improve the code quality, a linting software called ``detekt'',
+``[...] a static code analysis tool [...]''\cite{detekt2020} was used.
+
+Various libraries were used to help implement the application. Using
+the included build tool ``Gradle'' \cite{gradle} allowed easy management of 
+dependencies. The following libraries are worth mentioning: CameraX 
+\cite{camerax} is Androids newest Library allowing access to the Camera and 
+providing an easy way to analyse captured images. ``GPUImage'' \cite{gpuimage}
+allowed to set filters onto the preview image. A library called ``Nitrite''
+\cite{nitrite} allowed us to implement an NoSQL-Database
```
Comment (mac641) [inline chapters/implementation.tex]: Same as above

File: chapters/implementation.tex (hunk 5 above)
Comment (mac641) [inline chapters/implementation.tex]: Writing tense here, too - either use past (`was Android`) or change everything else to present
Androids -> Android's I think
newest -> latest
Library -> library

File: chapters/implementation.tex [hunk 10]
```diff
@@ -1,3 +1,54 @@
...
+Various libraries were used to help implement the application. Using
+the included build tool ``Gradle'' \cite{gradle} allowed easy management of 
+dependencies. The following libraries are worth mentioning: CameraX 
+\cite{camerax} is Androids newest Library allowing access to the Camera and 
+providing an easy way to analyse captured images. ``GPUImage'' \cite{gpuimage}
+allowed to set filters onto the preview image. A library called ``Nitrite''
+\cite{nitrite} allowed us to implement an NoSQL-Database 
+storing our measurement locally.
+
+For image analysis, we used the following libraries: For cascade-based face
```
Comment (mac641) [inline chapters/implementation.tex]: cascade-based -> cascade based

File: chapters/implementation.tex [hunk 11]
```diff
@@ -1,3 +1,54 @@
...
+detection, the official `OpenCV4Android'-Library was used\cite{opencv4android}.
+For machine learning based approaches, the official TFLite library was utilised.
```
Comment (mac641) [inline chapters/implementation.tex]: Three sentences starting with `For` here, I'd suggest rewriting them to provide a more fluent reading experience 😉

File: chapters/implementation.tex (hunk 8 above)
Comment (mac641) [inline chapters/implementation.tex]: Python 3-Script, running in an jupyter notebook, was utilised
Python 3-Script -> Python 3 script

**Review by mac641 — CHANGES_REQUESTED** [review]
Important: Change writing tense to present and make sure it's consistent throughout the entire text

File: chapters/implementation.tex [hunk 12]
```diff
@@ -1,3 +1,54 @@
...
+(IDE) for Android development, allowed faster development thanks to useful
+features like auto-completion. 
+
+An Android app is built into an `apk'-file. In order to reduce application
+size, soHappy differentiates between CPU architectures, and will
+build an apk for each architecture. As a minimum SDK version, version 23
+is selected. This means that the app works on Android 6 and above.
+
+In order to improve the code quality, a linting software called ``detekt'',
+``[...] a static code analysis tool [...]''\cite{detekt2020} was used.
```
Comment (limkokyew) [inline chapters/implementation.tex]: Not necessary, but you could provide a short description of what linting does, e. g. ensured coding conventions.

File: chapters/implementation.tex [hunk 13]
```diff
@@ -1,3 +1,54 @@
...
+detection use the official `OpenCV4Android'-Library \cite{opencv4android}.
+while machine learning based approaches utilise the official TFLite library.
+\cite{tensorflow}
+
+In order to train the CNN, which is explained in \ref{sec:methodology}, a
+Python 3 script running in an Jupyter notebook was utilised. 
+The Python TensorFlow library was used in combination with `Keras',
+which allows creating an model architecture easily. In order to prepare the
+training, test and validation data, the libraries `pandas', `pillow', 
+`mathplotlib' and `numpy' were used.
```
Comment (limkokyew) [inline chapters/implementation.tex]: "mathplotlib" -> "matplotlib"

File: chapters/implementation.tex [hunk 14]
```diff
@@ -1,3 +1,54 @@
...
+Thanks to the similarity with languages like Java and TypeScript, adapting
+Kotlin as programming language was relatively easy for the project team.
+
+Various language features of Kotlin helped in development, too. Instead of
+implementing the singleton pattern, Kotlin allows to create an single object
+per runtime by using the \texttt{object}-Keyword.
+
+Some design choices of the app were slowed down by Kotlin's language features.
+As all Kotlin classes inherit from \texttt{Any}, using features of Java's 
+\texttt{Object}-class are not natively available. This made using the methods
```
Comment (mac641) [inline chapters/implementation.tex]: I think this is still a Java issue, therefore: `made` -> `makes`

**Review by limkokyew — APPROVED** [review]
Good work! Just one suggestion and a minor typo. You have my approval nonetheless.

File: chapters/implementation.tex [hunk 15]
```diff
@@ -1,3 +1,54 @@
...
+\texttt{.wait()} and \texttt{.notify()}, which are used for synchronisation,
+more difficult.
+
+``Android Studio'', known as the official integrated development environment 
+(IDE) for Android development, allowed faster development thanks to useful
```
Comment (mac641) [inline chapters/implementation.tex]: `allowed` -> `allows`

File: chapters/implementation.tex (hunk 12 above)
Comment (mac641) [inline chapters/implementation.tex]: `was` -> `is`

File: chapters/implementation.tex [hunk 16]
```diff
@@ -1,3 +1,54 @@
...
+
+An Android app is built into an `apk'-file. In order to reduce application
+size, soHappy differentiates between CPU architectures, and will
+build an apk for each architecture. As a minimum SDK version, version 23
+is selected. This means that the app works on Android 6 and above.
+
+In order to improve the code quality, a linting software called ``detekt'',
+``[...] a static code analysis tool [...]''\cite{detekt2020} was used.
+
+Various libraries were used to help implement the application. Using
```
Comment (mac641) [inline chapters/implementation.tex]: `were` -> `are`

File: chapters/implementation.tex [hunk 17]
```diff
@@ -1,3 +1,54 @@
...
+the included build tool ``Gradle'' \cite{gradle} allowed easy management of
```
Comment (mac641) [inline chapters/implementation.tex]: `allowed` -> `allows`

File: chapters/implementation.tex [hunk 18]
```diff
@@ -1,3 +1,54 @@
...
+build an apk for each architecture. As a minimum SDK version, version 23
+is selected. This means that the app works on Android 6 and above.
+
+In order to improve the code quality, a linting software called ``detekt'',
+``[...] a static code analysis tool [...]''\cite{detekt2020} was used.
+
+Various libraries were used to help implement the application. Using
+the included build tool ``Gradle'' \cite{gradle} allowed easy management of 
+dependencies. The following libraries are worth mentioning: CameraX 
+\cite{camerax} is Android's newest library allowing access to the camera and
```
Comment (mac641) [inline chapters/implementation.tex]: `newest` -> `latest` is the better fitting word in this case, I think

File: chapters/implementation.tex [hunk 19]
```diff
@@ -1,3 +1,54 @@
...
+
+In order to improve the code quality, a linting software called ``detekt'',
+``[...] a static code analysis tool [...]''\cite{detekt2020} was used.
+
+Various libraries were used to help implement the application. Using
+the included build tool ``Gradle'' \cite{gradle} allowed easy management of 
+dependencies. The following libraries are worth mentioning: CameraX 
+\cite{camerax} is Android's newest library allowing access to the camera and 
+providing an easy way to analyse captured images. ``GPUImage'' \cite{gpuimage}
+allowed to apply filters onto the preview image. A library called ``Nitrite''
```
Comment (mac641) [inline chapters/implementation.tex]: `allowed` -> `allows`

File: chapters/implementation.tex [hunk 20]
```diff
@@ -1,3 +1,54 @@
...
+\cite{nitrite} allows to implement an NoSQL-Database storing our measurement 
+locally.
+
+For image analysis, we used the following libraries: Cascade based face
```
Comment (mac641) [inline chapters/implementation.tex]: `used` -> `use`

Comment (mac641) [reply chapters/implementation.tex]: By `Cascade`, you mean `Haar Cascade classifiers`?! 😅

File: chapters/implementation.tex [hunk 21]
```diff
@@ -1,3 +1,54 @@
...
+detection use the official `OpenCV4Android'-Library \cite{opencv4android}.
+while machine learning based approaches utilise the official TFLite library.
+\cite{tensorflow}
+
+In order to train the CNN, which is explained in \ref{sec:methodology}, a
+Python 3 script running in an Jupyter notebook was utilised.
```
Comment (mac641) [inline chapters/implementation.tex]: `was utilised` -> `is used`
utilise is somewhat a wrong fit here, I think

File: chapters/implementation.tex [hunk 22]
```diff
@@ -1,3 +1,54 @@
...
+locally.
+
+For image analysis, we used the following libraries: Cascade based face
+detection use the official `OpenCV4Android'-Library \cite{opencv4android}.
+while machine learning based approaches utilise the official TFLite library.
+\cite{tensorflow}
+
+In order to train the CNN, which is explained in \ref{sec:methodology}, a
+Python 3 script running in an Jupyter notebook was utilised. 
+The Python TensorFlow library was used in combination with `Keras',
```
Comment (mac641) [inline chapters/implementation.tex]: `was used` -> `is used`

File: chapters/implementation.tex (hunk 13 above)
Comment (mac641) [inline chapters/implementation.tex]: `were used` -> `are used`

**Review by mac641 — CHANGES_REQUESTED** [review]

**Review by diehmlukas — APPROVED** [review]

**Review by mac641 — APPROVED** [review]

**Pull request merged**